### PR TITLE
Implement ChakraRuntime::getPropertyNames() properly in native.

### PR DESCRIPTION
**Pull request by hansenyy**
In #3514, we modified ChakraRuntime::getPropertyNames() so that it returns all non-enumerable properties along the property chain (this matches the behavior of HermesRuntime.) However, in doing so, we switched its implementation from native to JS. This introduced a massive performance regression, and we had to revert back to the old implementation (see #3527.)

This PR attempts to correct the native implementation of ChakraRuntime::getPropertyNames(). The new implementation is still slower due to the checking of enumerability.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/3568)

### CODE REVIEW

File: vnext/JSI/Shared/ChakraRuntime.h [hunk 1]
```diff
@@ -248,15 +239,35 @@ class ChakraRuntime : public facebook::jsi::Runtime {
   ChakraObjectRef ToChakraObjectRef(const facebook::jsi::Value &value);
   std::vector<ChakraObjectRef> ToChakraObjectRefs(const facebook::jsi::Value *value, size_t count);
 
-  // Host function and host object helpers
+  // Convenience functions for property access.
+  ChakraObjectRef GetProperty(const ChakraObjectRef &obj, const ChakraObjectRef &id);
+  inline ChakraObjectRef GetProperty(const ChakraObjectRef &obj, const char *const name) {
```
Comment (NikoAri) [inline vnext/JSI/Shared/ChakraRuntime.h]: >  i [](start = 0, length = 3)

(nit) can you add empty line here for readability?

File: vnext/JSI/Shared/ChakraRuntime.h [hunk 2]
```diff
@@ -248,15 +239,35 @@ class ChakraRuntime : public facebook::jsi::Runtime {
   ChakraObjectRef ToChakraObjectRef(const facebook::jsi::Value &value);
   std::vector<ChakraObjectRef> ToChakraObjectRefs(const facebook::jsi::Value *value, size_t count);
 
-  // Host function and host object helpers
+  // Convenience functions for property access.
+  ChakraObjectRef GetProperty(const ChakraObjectRef &obj, const ChakraObjectRef &id);
```
Comment (NikoAri) [inline vnext/JSI/Shared/ChakraRuntime.h]: >id [](start = 81, length = 2)

(nit) just curious, why in one function this is "id" and in another it's "name" shouldn't they be consistent?

File: vnext/JSI/Shared/ChakraRuntime.h (hunk 1 above)
Comment (NikoAri) [inline vnext/JSI/Shared/ChakraRuntime.h]: >inline [](start = 2, length = 6)

(nit) is this needed? compilers do inlining anyways

File: vnext/JSI/Shared/ChakraRuntime.h [hunk 3]
```diff
@@ -248,15 +239,35 @@ class ChakraRuntime : public facebook::jsi::Runtime {
...
+  inline ChakraObjectRef GetProperty(const ChakraObjectRef &obj, const char *const name) {
+    return GetProperty(obj, GetChakraObjectRef(createPropNameIDFromAscii(name, strlen(name))));
```
Comment (NikoAri) [inline vnext/JSI/Shared/ChakraRuntime.h]: >createPropNameIDFromAscii [](start = 47, length = 25)

(nit) similarly, use "id" or "name" consistently.

File: vnext/JSI/Shared/ChakraRuntime.cpp [hunk 4]
```diff
@@ -376,25 +372,60 @@ bool ChakraRuntime::isHostFunction(const facebook::jsi::Function &obj) const {
...
+  // chain as soon as we hit Object.prototype. However, we still need to check
+  // for null here, as one can create an Object with no prototype through
+  // Object.create(null).
+  while (!CompareJsValues(currentObject, objectPrototype) && !CompareJsValues(currentObject, jsNull)) {
+    JsValueRef propNamesRef = JS_INVALID_REFERENCE;
+    VerifyJsErrorElseThrow(JsGetOwnPropertyNames(currentObject, &propNamesRef));
+    ChakraObjectRef propNames(propNamesRef);
 
-  VerifyJsErrorElseThrow(JsNumberToInt(countRef, &count));
+    int numPropNames = ToInteger(GetProperty(propNames, "length"));
```
Comment (NikoAri) [inline vnext/JSI/Shared/ChakraRuntime.cpp]: >numPropNames [](start = 8, length = 12)

(nit) propNamesLength? or propNamesSize?

File: vnext/JSI/Shared/ChakraRuntime.cpp [hunk 5]
```diff
@@ -376,25 +372,60 @@ bool ChakraRuntime::isHostFunction(const facebook::jsi::Function &obj) const {
...
 
-  auto result = createArray(count);
-  for (int i = 0; i < count; i++) {
-    JsValueRef index;
-    VerifyJsErrorElseThrow(JsIntToNumber(i, &index));
-    JsValueRef propertyName;
-    VerifyJsErrorElseThrow(JsGetIndexedProperty(propertyNamesArrayRef, index, &propertyName));
+    for (int i = 0; i < numPropNames; ++i) {
+      JsValueRef propNameRef = JS_INVALID_REFERENCE;
+      VerifyJsErrorElseThrow(JsGetIndexedProperty(propNames, ToJsNumber(i), &propNameRef));
```
Comment (NikoAri) [inline vnext/JSI/Shared/ChakraRuntime.cpp]: >  [](start = 75, length = 1)

(nit) /*out*/

File: vnext/JSI/Shared/ChakraRuntime.cpp [hunk 6]
```diff
@@ -376,25 +372,60 @@ bool ChakraRuntime::isHostFunction(const facebook::jsi::Function &obj) const {
...
-  VerifyJsErrorElseThrow(JsNumberToInt(countRef, &count));
+    int numPropNames = ToInteger(GetProperty(propNames, "length"));
 
-  auto result = createArray(count);
-  for (int i = 0; i < count; i++) {
-    JsValueRef index;
-    VerifyJsErrorElseThrow(JsIntToNumber(i, &index));
-    JsValueRef propertyName;
-    VerifyJsErrorElseThrow(JsGetIndexedProperty(propertyNamesArrayRef, index, &propertyName));
+    for (int i = 0; i < numPropNames; ++i) {
```
Comment (NikoAri) [inline vnext/JSI/Shared/ChakraRuntime.cpp]: >for [](start = 4, length = 3)

(nit) can you please add a small comment describing what this loop does? We are getting propNames on "currentObject", but then we are switching "currentObject" inside of loop.

File: vnext/JSI/Shared/ChakraRuntime.cpp [hunk 7]
```diff
@@ -376,25 +372,60 @@ bool ChakraRuntime::isHostFunction(const facebook::jsi::Function &obj) const {
...
+        enumerablePropNames.emplace_back(propName);
+      }
+    }
+
+    JsValueRef prototype = JS_INVALID_REFERENCE;
+    VerifyJsErrorElseThrow(JsGetPrototype(currentObject, &prototype));
+    currentObject = ChakraObjectRef(prototype);
+  }
+
+  size_t numEnumerablePropNames = enumerablePropNames.size();
```
Comment (NikoAri) [inline vnext/JSI/Shared/ChakraRuntime.cpp]: >numEnumerablePropNames [](start = 9, length = 22)

(nit) enumerablePropNamesSize or enumerablePropNamesLength

File: vnext/JSI/Shared/ChakraRuntime.cpp [hunk 8]
```diff
@@ -425,12 +456,26 @@ facebook::jsi::Array ChakraRuntime::createArray(size_t length) {
 
 size_t ChakraRuntime::size(const facebook::jsi::Array &arr) {
   assert(isArray(arr));
-  return static_cast<size_t>(GetProperty(arr, "length").asNumber());
+
+  int result = ToInteger(GetProperty(GetChakraObjectRef(arr), "length"));
+
+  if (result < 0 || result > SIZE_MAX) {
```
Comment (NikoAri) [inline vnext/JSI/Shared/ChakraRuntime.cpp]: >SIZE_MAX [](start = 29, length = 8)

What does this resolve to?

Comment (NikoAri) [inline vnext/JSI/Shared/ChakraRuntime.cpp]: >SIZE_MAX [](start = 29, length = 8)

use std::numeric_limits<T>::max instead

File: vnext/JSI/Shared/ChakraRuntime.cpp [hunk 9]
```diff
@@ -425,12 +456,26 @@ facebook::jsi::Array ChakraRuntime::createArray(size_t length) {
...
+    throw facebook::jsi::JSINativeException("Invalid JS array length detected.");
+  }
```
Comment (NikoAri) [inline vnext/JSI/Shared/ChakraRuntime.cpp]: (nit) can we replace this with something like VerifyElseThrow(result < 0 || result > SIZE_MAX, "Invalid JS array length detected.")

File: vnext/JSI/Shared/ChakraRuntime.h (hunk 2 above)
Comment (hansenyy) [reply vnext/JSI/Shared/ChakraRuntime.h]: The difference is intentional as a JS property ID can either be backed by a Symbol or a name (a.k.a. string)

---
In reply to: [341261195](https://github.com/microsoft/react-native-windows/pull/3568#discussion_r341261195) [](ancestors = 341261195)

File: vnext/JSI/Shared/ChakraRuntime.h (hunk 1 above)
Comment (hansenyy) [reply vnext/JSI/Shared/ChakraRuntime.h]: Same with previous comment.

---
In reply to: [341260719](https://github.com/microsoft/react-native-windows/pull/3568#discussion_r341260719) [](ancestors = 341260719)

Comment (hansenyy) [reply vnext/JSI/Shared/ChakraRuntime.h]: Yea the compiler should automatically do inlining here, but I like the explicitness.

---
In reply to: [341261679](https://github.com/microsoft/react-native-windows/pull/3568#discussion_r341261679) [](ancestors = 341261679)

File: vnext/JSI/Shared/ChakraRuntime.h (hunk 3 above)
Comment (hansenyy) [reply vnext/JSI/Shared/ChakraRuntime.h]: Same as previous comment.

---
In reply to: [341262364](https://github.com/microsoft/react-native-windows/pull/3568#discussion_r341262364) [](ancestors = 341262364)

File: vnext/JSI/Shared/ChakraRuntime.cpp [hunk 10]
```diff
@@ -376,25 +372,60 @@ bool ChakraRuntime::isHostFunction(const facebook::jsi::Function &obj) const {
...
+  ChakraObjectRef objectConstructor = GetProperty(GetChakraObjectRef(global()), "Object");
+
+  // Handle to the Object.prototype Object.
+  ChakraObjectRef objectPrototype = GetProperty(objectConstructor, "prototype");
+
+  // Handle to the Object.prototype.propertyIsEnumerable() Function.
+  ChakraObjectRef objectPrototypePropertyIsEnumerable = GetProperty(objectPrototype, "propertyIsEnumerable");
+
+  std::vector<ChakraObjectRef> enumerablePropNames{};
+  ChakraObjectRef currentObject = GetChakraObjectRef(object);
```
[the remark this answers is not in the document]
Comment (hansenyy) [reply vnext/JSI/Shared/ChakraRuntime.cpp]: I'll rename currentObjec to currentObjectOnPrototypeChain.

---
In reply to: [341264449](https://github.com/microsoft/react-native-windows/pull/3568#discussion_r341264449) [](ancestors = 341264449)

File: vnext/JSI/Shared/ChakraRuntime.cpp [hunk 11]
```diff
@@ -376,25 +372,60 @@ bool ChakraRuntime::isHostFunction(const facebook::jsi::Function &obj) const {
...
-  JsValueRef countRef;
-  VerifyJsErrorElseThrow(JsGetProperty(propertyNamesArrayRef, propertyId, &countRef));
-  int count;
+  // We have a small optimization here where we stop traversing the prototype
+  // chain as soon as we hit Object.prototype. However, we still need to check
+  // for null here, as one can create an Object with no prototype through
+  // Object.create(null).
+  while (!CompareJsValues(currentObject, objectPrototype) && !CompareJsValues(currentObject, jsNull)) {
+    JsValueRef propNamesRef = JS_INVALID_REFERENCE;
+    VerifyJsErrorElseThrow(JsGetOwnPropertyNames(currentObject, &propNamesRef));
```
[the remark this answers is not in the document]
Comment (hansenyy) [reply vnext/JSI/Shared/ChakraRuntime.cpp]: We will need to add this to all sites where Chakra APIs are called. Punting this work for now.

---
In reply to: [341265034](https://github.com/microsoft/react-native-windows/pull/3568#discussion_r341265034) [](ancestors = 341265034)

File: vnext/JSI/Shared/ChakraRuntime.cpp (hunk 5 above)
Comment (hansenyy) [reply vnext/JSI/Shared/ChakraRuntime.cpp]: See other comment.

---
In reply to: [341265762](https://github.com/microsoft/react-native-windows/pull/3568#discussion_r341265762) [](ancestors = 341265762)

File: vnext/JSI/Shared/ChakraRuntime.cpp (hunk 9 above)
Comment (hansenyy) [reply vnext/JSI/Shared/ChakraRuntime.cpp]: There are other places that might benefit from a similar pattern. I think we should put this on hold till when we work on ensuring correct exception propagation.

---
In reply to: [341273515](https://github.com/microsoft/react-native-windows/pull/3568#discussion_r341273515) [](ancestors = 341273515)

**Pull request merged**